xcode/Subconscious/Shared/Services/GatewayProvisioningService.swift

```diff
@@ -0,0 +1,230 @@
+                "Could not read response"
+            )
+        }
+        
+        guard response.success else {
+            let responseBody = try self.jsonDecoder.decode(
+                ProvisionGatewayErrorResponse.self,
+                from: data
+            )
+            
+            if let error = responseBody.error {
+                throw GatewayProvisioningServiceError.failedToProvisionGateway(
+                    "HTTP \(response.statusCode): \(error)"
+                )
+            } else {
+                throw GatewayProvisioningServiceError.failedToProvisionGateway(
+                    "Unexpected status code \(response.statusCode)"
+                )
+            }
+        }
+        
+        return try self.jsonDecoder.decode(
+            ProvisionGatewayResponse.self,
+            from: data
+        )
+    }
+    
+    private func checkGatewayProvisioningStatus(
+        gatewayId: String
+    ) async throws -> ProvisionGatewayStatusResponse {
+        var request = URLRequest(url: Self.provisionGatewayEndpoint.appending(path: gatewayId))
+        request.httpMethod = "GET"
+        
+        let (data, response) = try await URLSession.shared.data(for: request)
+        
+        guard let response = response as? HTTPURLResponse else {
+            throw GatewayProvisioningServiceError.failedToCheckGatewayProvisioningStatus(
+                "Could not read response"
+            )
+        }
+        
+        guard response.success else {
+            throw GatewayProvisioningServiceError.failedToCheckGatewayProvisioningStatus(
+                "Unexpected status code \(response.statusCode)"
+            )
+        }
+        
+        return try self.jsonDecoder.decode(
+            ProvisionGatewayStatusResponse.self,
+            from: data
+        )
+    }
+    
+    func waitForGatewayProvisioning(
+        gatewayId: String,
+        maxAttempts: Int,
+        attempts: Int = 0
+    ) async throws -> URL? {
+        do {
+            let res = try await self.checkGatewayProvisioningStatus(gatewayId: gatewayId)
+            
+            guard res.status.lowercased() == "active" else {
+                throw GatewayProvisioningServiceError.gatewayIsNotReady
+            }
+            
+            // Validate response payload
+            guard let did = res.did,
+                  let _ = Did(did),
+                  let url = res.url,
+                  let url = URL(string: url) else {
+                throw GatewayProvisioningServiceError.invalidStatusResponse(res)
+            }
+            
+            return url
+        } catch {
+            let attempts = attempts + 1
+            guard attempts < maxAttempts else {
+                return nil
+            }
+            
+            Self.logger.log("""
+            Waiting to re-check provisioning status, \
+            attempt \(attempts) of \(maxAttempts)
+            """)
+            
+            sleep(UInt32(attempts))
```

Comment:
Simple linear backoff, we could easily switch to exponential if this is too eager.

Reply:
Note: extract this waiting loop somewhere else, we will want to re-use it.

Do 2^n with a configurable maximum wait time.